src/Usos/Server.php

```diff
@@ -0,0 +1,162 @@
+    {
+        return $this->getDomain().'/services/oauth/request_token';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function urlAuthorization()
+    {
+        return $this->getDomain().'/services/oauth/authorize';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function urlTokenCredentials()
+    {
+        return $this->getDomain().'/services/oauth/access_token';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function urlUserDetails()
+    {
+        return $this->getDomain().'/services/users/user'.'?fields=id|first_name|last_name|email|photo_urls';
+    }
+
+    private function getDomain()
+    {
+        return config('services.usos.domain');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function userDetails($data, TokenCredentials $tokenCredentials)
+    {
+        $user = new User();
+        $user->id = $data['id'];
+        $user->nickname = $data['first_name'].' '.$data['last_name'];
+        $user->name = $data['first_name'].' '.$data['last_name'];
+        $user->avatar = $data['photo_urls']['50x50'] ?? null;
+        $user->email = $data['email'] ?? null;
+
+        $used = ['id', 'first_name', 'last_name', 'email'];
+
+        $user->extra = array_diff_key($data, array_flip($used));
+
+        return $user;
+    }
+
+    /**
+     * Generate the OAuth protocol header for a temporary credentials
+     * request, based on the URI.
+     *
+     * @param string $uri
+     *
+     * @return string
+     */
+    protected function temporaryCredentialsProtocolHeader($uri)
+    {
+        $parameters = array_merge($this->baseProtocolParameters(), array_merge([
+            'oauth_callback' => $this->clientCredentials->getCallbackUri(),
+
+        ], ['scopes' => implode('|', $this->scopes)]));
+
+        $parameters['oauth_signature'] = $this->signature->sign($uri, $parameters, 'POST');
+
+        return $this->normalizeProtocolParameters($parameters);
+    }
+
+    public function getTemporaryCredentials()
+    {
+        $uri = $this->urlTemporaryCredentials();
+
+        $client = $this->createHttpClient();
+
+        $formParams = [
+            'scopes' => implode('|', $this->scopes),
+        ];
+
+        $header = $this->temporaryCredentialsProtocolHeader($uri);
+        $authorizationHeader = ['Authorization' => $header];
+        $headers = $this->buildHttpClientHeaders($authorizationHeader);
+
+        try {
+            $response = $client->post($uri, [
+                'headers'     => $headers,
+                'form_params' => $formParams,
+            ]);
+
+            return $this->createTemporaryCredentials((string) $response->getBody());
+        } catch (BadResponseException $e) {
+            $this->handleTemporaryCredentialsBadResponse($e);
+        }
+
+        throw new CredentialsException('Failed to get temporary credentials');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function userUid($data, TokenCredentials $tokenCredentials)
```

Comment:
What are these functions returned for?

Reply:
Because of Server class extends BaseServer class, which based on an abstract Server class of League/OAuth1-client with abstract userUid method, the definition of this function is necessary.

According to the docs of league/oauth1-client, this function must return only unique authenticated user's id from list of user's data, which was returned from the user details URL. In this case, it takes user's unique ID from USOS data list, ([services/users/user](https://apps.usos.edu.pl/developers/api/services/users/#user)) and returns it.

Reply:
All good, I have not reviewed many v1 providers 😅